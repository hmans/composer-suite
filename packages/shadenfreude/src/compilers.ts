import { RenderCallback } from "@react-three/fiber"
import { IUniform } from "three"
import { variablesToNodes } from "./factories"
import { formatValue } from "./formatters"
import { isVariable, ShaderNode, Value, Variable } from "./types"

type Program = "vertex" | "fragment"

export function compileVariableValue(value: Value): string {
  return isVariable(value) ? value.name : formatValue(value)
}

export function compileVariable(variable: Variable) {
  const valueString =
    variable.qualifier !== "uniform" && variable.value !== undefined
      ? ` = ${compileVariableValue(variable.value)}`
      : ""

  return `${variable.qualifier ?? ""} ${variable.type} ${
    variable.name
  }${valueString};`
}

function nodeTitle(node: ShaderNode) {
  return `/** Node: ${node.name} **/`
}

function dependencies(node: ShaderNode, deps = new Array<ShaderNode>()) {
  for (const variable of Object.values(node.inputs)) {
    if (isVariable(variable.value)) {
      /* get dependency */
      const dependency = variablesToNodes.get(variable.value)!
      if (!dependency) throw new Error("Dependency not found")

      /* If we haven't seen this dependency yet, invoke the callback */
      if (!deps.includes(dependency)) {
        dependencies(dependency, deps)
      }
    }
  }

  deps.push(node)
  return deps
}

function compileHeader(node: ShaderNode, program: Program) {
  const header = (n: ShaderNode) => `
    ${nodeTitle(n)}

    ${Object.values(n.uniforms)
      .map((variable) => compileVariable({ ...variable, qualifier: "uniform" }))
      .join("\n")}

    ${Object.values(n.varyings)
      .map((variable) => compileVariable({ ...variable, qualifier: "varying" }))
      .join("\n")}

    ${n[program].header ?? ""}
  `

  return dependencies(node)
    .map(header)
    .join("\n\n\n")
}

function compileBody(node: ShaderNode, program: Program) {
  return dependencies(node)
    .map(
      (node) => `
        ${nodeTitle(node)}

        ${Object.entries(node.outputs)
          .map(([_, variable]) =>
            compileVariable({ ...variable, value: undefined })
          )
          .join("")}

        {
          /* Varying References */
          ${Object.entries(node.varyings)
            .map(([name, variable]) =>
              compileVariable({ ...variable, name, value: variable.name })
            )
            .join("")}

          /* Inputs */
          ${Object.entries(node.inputs)
            .map(([name, variable]) => compileVariable({ ...variable, name }))
            .join("")}

          /* Outputs */
          ${Object.entries(node.outputs)
            .map(([name, variable]) => compileVariable({ ...variable, name }))
            .join("")}

          /* Code */
          ${node[program].body ?? ""}

          /* Update globals */
          ${Object.entries(node.outputs)
            .map(([name, variable]) => `${variable.name} = ${name};`)
            .join("\n")}
          ${
            program === "vertex"
              ? Object.entries(node.varyings)
                  .map(([name, variable]) => `${variable.name} = ${name};`)
                  .join("")
              : ""
          }
          }
      `
    )
    .join("\n\n\n")
}

export function compileShader(root: ShaderNode) {
  /* Build uniforms */
  const uniforms = {} as Record<string, IUniform>

  dependencies(root).forEach((node) => {
    Object.entries(node.uniforms).forEach(([name, variable]) => {
      uniforms[variable.name] = variable as IUniform
    })
  })

  /* Build callbacks */
  const callbacks = dependencies(root).map((node) => node.update)

  const update: RenderCallback = (...args) => {
    callbacks.forEach((callback) => callback?.(...args))
  }

  /* Build vertex shader program */
  const vertexShader = `
    /*** VERTEX SHADER ***/

    ${compileHeader(root, "vertex")}

    void main() {
      ${compileBody(root, "vertex")}
    }`

  /* Build fragment shader program */
  const fragmentShader = `
    /*** FRAGMENT SHADER ***/

    ${compileHeader(root, "fragment")}

    void main() {
      ${compileBody(root, "fragment")}
    }`

  return {
    vertexShader,
    fragmentShader,
    uniforms,
    update
  }
}
