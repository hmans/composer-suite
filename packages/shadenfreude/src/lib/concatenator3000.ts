export type Parts = any[]

const compact = (p: any) => !!p

const indent = (p: string) => "  " + p

export const block = (...parts: Parts): Parts => {
  const flattened = flatten(parts)
  return flattened.length > 0 ? ["{", flattened.map(indent), "}"] : []
}

export const concatenate = (...parts: Parts) => flatten(...parts).join("\n")

export const line = (...parts: Parts) => flatten(...parts).join(" ")

export const flatten = (...parts: Parts): Parts =>
  parts
    .filter(compact)
    .map((p) => (Array.isArray(p) ? flatten(...p) : p))
    .flat()

export const comment = (...parts: Parts) => line("/*", ...parts, "*/")

export const statement = (...parts: Parts) => line(...parts) + ";"

export const assignment = (left: string, right: string) =>
  statement(left, "=", right)

export const identifier = (...parts: Parts) =>
  parts
    .flat()
    .filter(compact)
    .join("_")
    .replace(/_{2,}/g, "_")

export const sluggify = (s: string) =>
  s.replace(/[^a-zA-Z0-9]/g, "_").replace(/_{2,}/g, "_")
