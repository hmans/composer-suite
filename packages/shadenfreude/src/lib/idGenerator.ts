export default (initial = 0) => () => ++initial
