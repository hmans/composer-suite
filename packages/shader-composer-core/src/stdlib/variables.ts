import { $, Expression } from "../expressions"
import { frame } from "../ticker"
import { GLSLType, injectAPI, JSTypes, Unit, UnitConfig } from "../units"
import { $localToViewSpace, $localToWorldSpace } from "./spaces"
import { Bool, Int, Mat4, Vec2, Vec3 } from "./values"

/**
 * Returns the current fragment's on-screen coordinate.
 */
export const FragmentCoordinate = Vec2($`gl_FragCoord.xy`, {
  name: "Fragment Coordinate",
  only: "fragment"
})

/**
 * In instanced rendering, will return the instance ID.
 * Wraps the `gl_InstanceID` GLSL built-in.
 */
export const InstanceID = Int($`gl_InstanceID`, {
  name: "Instance ID",
  varying: "flat"
})

/**
 * Returns the verte ID.
 * Wraps the `gl_VertexID` GLSL built-in.
 *
 * Note: available in vertex shader only!
 */
export const VertexID = Int($`gl_VertexID`, {
  name: "Vertex ID",
  only: "vertex"
})

export const CameraPosition = Vec3($`cameraPosition`, {
  name: "Camera Position"
})

export const ViewMatrix = Mat4($`viewMatrix`, {
  name: "View Matrix"
})

export const ModelMatrix = Mat4($`modelMatrix`, {
  name: "Model Matrix"
})

export const ModelViewMatrix = Mat4($`modelViewMatrix`, {
  name: "ModelView Matrix"
})

export const NormalMatrix = Mat4($`normalMatrix`, {
  name: "Normal Matrix"
})

export const ProjectionMatrix = Mat4($`projectionMatrix`, {
  name: "Projection Matrix"
})

/**
 * Returns true if instanced rendering is enabled, false if it is not.
 */
export const UsingInstancing = Bool($`
  #ifdef USE_INSTANCING
    true
  #else
    false
  #endif
`)

/**
 * Returns the instance matrix. Please note that this is only available when
 * instanced rendering is enabled.
 */
export const InstanceMatrix = Mat4($`instanceMatrix`, {
  name: "Instance Matrix",
  only: "vertex"
})

export const UV = Vec2($`uv`, {
  name: "UV",
  varying: true
})

const Vec3WithSpaceConversions = (expr: Expression, name: string) =>
  injectAPI(Vec3(expr, { name, varying: true }), (unit) => ({
    get world() {
      return Vec3($localToWorldSpace(unit), {
        varying: true,
        name: `${name} (World Space)`
      })
    },

    get view() {
      return Vec3($localToViewSpace(unit), {
        varying: true,
        name: `${name} (View Space)`
      })
    }
  }))

export const VertexPosition = Vec3WithSpaceConversions(
  $`position`,
  "Vertex Position"
)
export const VertexNormal = Vec3WithSpaceConversions($`normal`, "Vertex Normal")

export const ViewDirection = Vec3(
  $`vec3(-${ViewMatrix}[0][2], -${ViewMatrix}[1][2], -${ViewMatrix}[2][2])`,
  { varying: true, name: "View Direction" }
)

export const IsFrontFacing = Bool($`gl_FrontFacing`, { only: "fragment" })

export const Attribute = <T extends GLSLType>(type: T, name: string) =>
  Unit(type, $`${name}`, {
    name: `Attribute: ${name}`,
    varying: true,
    vertex: {
      header: $`attribute ${type} ${name};`
    }
  })

export type UniformUnit<
  T extends GLSLType,
  J extends JSTypes[T] = JSTypes[T]
> = Unit<T> & {
  value: J
}

export const UniformUnit = <T extends GLSLType, J extends JSTypes[T]>(
  type: T,
  initialValue: J,
  extras?: Partial<UnitConfig<T>>
): UniformUnit<T, J> => {
  const uniform = { value: initialValue }

  /* Create the actual unit that represents the uniform. */
  const unit = Unit(type, undefined, {
    name: `Uniform (${type})`,
    ...extras,
    uniform
  })

  /* Return the unit with some API bits mixed in. */
  return injectAPI(unit, () => ({
    set value(v: J) {
      uniform.value = v
    },

    get value(): J {
      return uniform.value
    }
  }))
}

/**
 * Provides a uniform unit holding a representation of time. The time value
 * stored is not an absolute time, so multiple instances of this unit will not
 * be synchronized. If you require synchronization, please either reuse the
 * same instance of this unit, or use `GlobalTime` instead.
 *
 * @param initial The initial time value to start with. (Default: 0)
 */
export const Time = (initial: number = 0) => {
  const uniform = UniformUnit("float", initial, {
    name: "Time Uniform",

    update: (dt) => {
      uniform.value += dt
    }
  })

  return uniform
}

/**
 * A global time uniform unit that can be safely used across multiple shaders,
 * wherever synchronization is required,
 * and as a default value for `time` inputs of other unit implementations, to prevent
 * shaders from being spammed by multiple uniforms all holding different
 * representations of time.
 */
export const GlobalTime = Time()

/**
 * An absolute time value, in seconds, that remains stable during the duration
 * of the currently rendered frame. This is useful for creating effects that
 * require accurate synchronization across multiple shaders.
 */
export const FrameTime = UniformUnit("float", frame.time, {
  name: "Frame Time",

  update: () => {
    FrameTime.value = frame.time
  }
})

/**
 * An absolute count of the number of frames that have been rendered so far.
 * This is useful for creating effects that require accurate synchronization
 * across multiple shaders, and prefer to use an integer frame count instead
 * of a floating point time value like {@link FrameTime} provides.
 */
export const FrameCount = UniformUnit("int", frame.count, {
  name: "Frame Count",

  update: () => {
    FrameCount.value = frame.count
  }
})

export const Resolution = UniformUnit(
  "vec2",
  { x: 0, y: 0 },
  {
    name: "Current Render Resolution",

    update: () => {
      "Please customize the Resolution unit with the relevant code for your environment."
    }
  }
)

export const CameraNear = UniformUnit("float", 0 as number, {
  name: "Camera Near Plane",

  update: () => {
    console.warn(
      "Please customize the CameraNear unit with the relevant code for your environment."
    )
  }
})

export const CameraFar = UniformUnit("float", 0 as number, {
  name: "Camera Far Plane",

  update: () => {
    console.warn(
      "Please customize the CameraNear unit with the relevant code for your environment."
    )
  }
})

export const ScreenUV = Vec2($`${FragmentCoordinate} / ${Resolution}`, {
  name: "Screen UV"
})
