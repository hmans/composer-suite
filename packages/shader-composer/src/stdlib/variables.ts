import { PerspectiveCamera, Vector2 } from "three"
import { $, Expression } from "../expressions"
import { GLSLType, injectAPI, Input, JSTypes, Unit, UnitConfig } from "../units"
import { ScaleAndOffset } from "./math"
import { $localToViewSpace, $localToWorldSpace } from "./spaces"
import { Bool, Int, Mat4, Vec2, Vec3 } from "./values"

/**
 * Returns the current fragment's on-screen coordinate.
 */
export const FragmentCoordinate = Vec2($`gl_FragCoord.xy`, {
  name: "Fragment Coordinate",
  only: "fragment"
})

/**
 * In instanced rendering, will return the instance ID.
 * Wraps the `gl_InstanceID` GLSL built-in.
 */
export const InstanceID = Int(0, {
  name: "Instance ID",
  varying: "flat"
})

/**
 * Returns the verte ID.
 * Wraps the `gl_VertexID` GLSL built-in.
 *
 * Note: available in vertex shader only!
 */
export const VertexID = Int($`gl_VertexID`, {
  name: "Vertex ID",
  only: "vertex"
})

export const CameraPosition = Vec3($`cameraPosition`, {
  name: "Camera Position"
})

export const ViewMatrix = Mat4($`viewMatrix`, {
  name: "View Matrix"
})

export const ModelMatrix = Mat4($`modelMatrix`, {
  name: "Model Matrix"
})

export const ModelViewMatrix = Mat4($`modelViewMatrix`, {
  name: "ModelView Matrix"
})

export const NormalMatrix = Mat4($`normalMatrix`, {
  name: "Normal Matrix"
})

export const ProjectionMatrix = Mat4($`projectionMatrix`, {
  name: "Projection Matrix"
})

/**
 * Returns true if instanced rendering is enabled, false if it is not.
 */
export const UsingInstancing = Bool($`
  #ifdef USE_INSTANCING
    true
  #else
    false
  #endif
`)

/**
 * Returns the instance matrix. Please note that this is only available when
 * instanced rendering is enabled.
 */
export const InstanceMatrix = Mat4($`instanceMatrix`, {
  name: "Instance Matrix",
  only: "vertex"
})

export const UV = Vec2($`uv`, {
  name: "UV",
  varying: true
})

const Vec3WithSpaceConversions = (expr: Expression, name: string) =>
  injectAPI(Vec3(expr, { name, varying: true }), (unit) => ({
    get world() {
      return Vec3($localToWorldSpace(unit), {
        varying: true,
        name: `${name} (World Space)`
      })
    },

    get view() {
      return Vec3($localToViewSpace(unit), {
        varying: true,
        name: `${name} (View Space)`
      })
    }
  }))

export const VertexPosition = Vec3WithSpaceConversions(
  $`position`,
  "Vertex Position"
)
export const VertexNormal = Vec3WithSpaceConversions($`normal`, "Vertex Normal")

export const ViewDirection = Vec3(
  $`vec3(-${ViewMatrix}[0][2], -${ViewMatrix}[1][2], -${ViewMatrix}[2][2])`,
  { varying: true, name: "View Direction" }
)

export const IsFrontFacing = Bool($`gl_FrontFacing`, { only: "fragment" })

export const Attribute = <T extends GLSLType>(type: T, name: string) =>
  Unit(type, $`${name}`, {
    name: `Attribute: ${name}`,
    varying: true,
    vertex: {
      header: $`attribute ${type} ${name};`
    }
  })

export type UniformUnit<
  T extends GLSLType,
  J extends JSTypes[T] = JSTypes[T]
> = Unit<T> & {
  value: J
}

export const UniformUnit = <T extends GLSLType, J extends JSTypes[T]>(
  type: T,
  initialValue: J,
  extras?: Partial<UnitConfig<T>>
): UniformUnit<T, J> => {
  const uniform = { value: initialValue }

  /* Create the actual unit that represents the uniform. */
  const unit = Unit(type, undefined, {
    name: `Uniform (${type})`,
    ...extras,
    uniform
  })

  /* Return the unit with some API bits mixed in. */
  return injectAPI(unit, () => ({
    set value(v: J) {
      uniform.value = v
    },

    get value(): J {
      return uniform.value
    }
  }))
}

/**
 * Provides a uniform unit holding a representation of time. The time value
 * stored is not an absolute time, so multiple instances of this unit will not
 * be synchronized. If you require synchronization, please either reuse the
 * same instance of this unit, or use `GlobalTime` instead.
 *
 * @param initial The initial time value to start with. (Default: 0)
 */
export const Time = (initial: number = 0) => {
  const uniform = UniformUnit("float", initial, {
    name: "Time Uniform",

    update: (dt) => {
      uniform.value += dt
    }
  })

  return uniform
}

/**
 * A global time uniform unit that can be safely used across multiple shaders,
 * wherever synchronization is required,
 * and as a default value for `time` inputs of other unit implementations, to prevent
 * shaders from being spammed by multiple uniforms all holding different
 * representations of time.
 */
export const GlobalTime = Time()

export const Resolution = UniformUnit("vec2", new Vector2(0, 0), {
  name: "Current Render Resolution",

  update: (dt, camera, scene, gl) => {
    Resolution.value.x = gl.domElement.width
    Resolution.value.y = gl.domElement.height
  }
})

export const CameraNear = UniformUnit("float", 0 as number, {
  name: "Camera Near Plane",

  update: (_, camera) => {
    if (camera instanceof PerspectiveCamera) {
      CameraNear.value = camera.near
    }
  }
})

export const CameraFar = UniformUnit("float", 0 as number, {
  name: "Camera Far Plane",

  update: (_, camera) => {
    if (camera instanceof PerspectiveCamera) {
      CameraFar.value = camera.far
    }
  }
})

export const ScreenUV = Vec2($`${FragmentCoordinate} / ${Resolution}`, {
  name: "Screen UV"
})
