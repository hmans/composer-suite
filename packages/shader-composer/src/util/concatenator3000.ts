export type Part = any

export const isPresent = (p: Part) => !!p

export const concatenate = (...parts: Part[]): string =>
	parts
		.flat(Infinity)
		.filter(isPresent)
		.join("\n")

export const identifier = (...parts: Part[]): string =>
	parts
		.filter(isPresent)
		.join("_")
		.replace(/_{2,}/g, "_")

export const statement = (...parts: Part[]): string =>
	parts.filter(isPresent).join(" ") + ";"

export const assignment = (...parts: Part[]): string => statement(parts.join(" = "))

export const block = (...parts: Part[]): Part[] => [
	"{",
	concatenate(parts)
		.split("\n")
		.map((line) => "  " + line)
		.join("\n"),
	"}"
]

export const sluggify = (s: string) => s.replace(/[^a-zA-Z0-9]/g, "_")
