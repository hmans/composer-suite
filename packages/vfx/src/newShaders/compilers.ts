import { formatValue } from "../shaders"
import { IVariable, Shader, VariableQualifier, Variables } from "./types"

function compileVariable(
  qualifier: VariableQualifier,
  name: string,
  { type, value }: IVariable
) {
  const valueString = value !== undefined ? ` = ${formatValue(value)}` : ""
  return `${qualifier} ${type} ${name}${valueString};`
}

function compileVariables(qualifier: VariableQualifier, variables: Variables) {
  return Object.entries(variables)
    .map(([name, variable]) => compileVariable(qualifier, name, variable))
    .join("\n")
}

function compileProgram(
  shader: Shader,
  program: "vertexShader" | "fragmentShader"
) {
  const {
    uniforms,
    varyings,
    [program]: { header, main }
  } = shader

  return `
    ${compileVariables("uniform", uniforms)}
    ${compileVariables("varying", varyings)}
    ${header}

    void main() {
      ${main}
    }`
}

export function compileMaterial(shader: Shader) {
  return {
    uniforms: shader.uniforms,
    vertexShader: compileProgram(shader, "vertexShader"),
    fragmentShader: compileProgram(shader, "fragmentShader")
  }
}
