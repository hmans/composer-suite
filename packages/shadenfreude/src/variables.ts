import { Color, Matrix3, Matrix4, Vector2, Vector3, Vector4 } from "three"
import { glslRepresentation } from "./glslRepresentation"
import { type } from "./glslType"
import { identifier } from "./lib/concatenator3000"
import idGenerator from "./lib/idGenerator"

export type GLSLType =
  | "bool"
  | "float"
  | "vec2"
  | "vec3"
  | "vec4"
  | "mat3"
  | "mat4"

export type JSTypes = {
  bool: boolean
  float: number
  vec2: Vector2
  vec3: Vector3 | Color
  vec4: Vector4
  mat3: Matrix3
  mat4: Matrix4
}

export type Value<T extends GLSLType = any> = string | JSTypes[T] | Variable<T>

export type Variable<T extends GLSLType = any> = {
  _: "Variable"
  id: number
  title: string
  name: string
  type: T
  value: Value<T>
  inputs: Record<string, Value>
  only?: "vertex" | "fragment"
  vertexHeader?: string
  vertexBody?: string
  fragmentHeader?: string
  fragmentBody?: string

  /*
  Some convenient helpers to directly apply math operations to variables.
  Do these increase complexity too much for some cute glue? Maybe, maybe not.
  */
  Add: (...operands: Value[]) => Variable<T>
  Subtract: (...operands: Value[]) => Variable<T>
  Multiply: (...operands: Value[]) => Variable<T>
  Divide: (...operands: Value[]) => Variable<T>
}

const nextAnonymousId = idGenerator()

const buildMultiInputs = (values: Value[]) =>
  values.reduce((acc, v, i) => ({ ...acc, [`m_${i}`]: v }), {})

export const Variable = <T extends GLSLType>(
  type: T,
  value: Value<T>,
  extras: Partial<Variable<T>> = {}
) => {
  const id = nextAnonymousId()

  const v: Variable<T> = {
    id,
    title: `Anonymous ${type} = ${glslRepresentation(value)}`,
    name: identifier("anonymous", id),
    inputs: {},
    ...extras,
    _: "Variable",
    type,
    value,

    /* Math helpers! */
    Add: (...operands: Value[]) => Add(v, ...operands),
    Subtract: (...operands: Value[]) => Subtract(v, ...operands),
    Multiply: (...operands: Value[]) => Multiply(v, ...operands),
    Divide: (...operands: Value[]) => Divide(v, ...operands)
  }

  return v
}

export function isVariable(v: any): v is Variable {
  return v && v._ === "Variable"
}

export function isType<T extends GLSLType>(v: any, t: T): v is Value<T> {
  return type(v) === t
}

/* Helpers */

const makeVariableHelper = <T extends GLSLType>(type: T) => (
  v: Value<T>,
  extras?: Partial<Variable<T>>
) => Variable(type, v, extras)

export const Float = makeVariableHelper("float")
export const Bool = makeVariableHelper("bool")
export const Vec2 = makeVariableHelper("vec2")
export const Vec3 = makeVariableHelper("vec3")
export const Vec4 = makeVariableHelper("vec4")
export const Mat3 = makeVariableHelper("mat3")
export const Mat4 = makeVariableHelper("mat4")

/*** OPERATORS ***

TODO: Move this somewhere else!

It used to be in maths.ts, which caused a circular dependency. Pretty much everything
was fine with this except for ts-jest, which objected strongly, and crashed.
*/

export const Operator = (title: string, operator: "+" | "-" | "*" | "/") => <
  T extends GLSLType
>(
  a: Value<T>,
  ...rest: Value[]
) => {
  const inputs = buildMultiInputs([a, ...rest])

  /* a + b + c + ... */
  const expression = Object.keys(inputs).join(operator)

  return Variable(type(a), expression, {
    title: `${title} (${type(a)})`,
    inputs
  })
}

export const Add = Operator("Add", "+")
export const Subtract = Operator("Subtract", "-")
export const Multiply = Operator("Multiply", "*")
export const Divide = Operator("Divide", "/")
