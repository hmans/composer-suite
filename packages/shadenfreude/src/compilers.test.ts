import { compileShader } from "./compilers"
import { assignment, statement } from "./lib/concatenator3000"
import { Bool, Float, Variable } from "./variables"

describe("compileShader", () => {
  it("returns a vertexShader", () => {
    const v = Variable("float", 1)
    const [shader] = compileShader(v)

    expect(shader.vertexShader).toMatchSnapshot()
  })

  it("returns a fragmentShader", () => {
    const v = Variable("float", 1)
    const [shader] = compileShader(v)

    expect(shader.fragmentShader).toMatchSnapshot()
  })

  it("includes the variable's chunks if it has them", () => {
    const v = Variable("float", 1)

    v._config.vertexHeader = statement("uniform float u_time")
    v._config.vertexBody = assignment(
      "gl_Position",
      "vec4(sin(u_time), 0.0, 0.0, 1.0)"
    )
    v._config.fragmentHeader = statement("uniform float u_time")
    v._config.fragmentBody = assignment(
      "gl_FragColor",
      "vec4(cos(u_time), 0.0, 0.0, 1.0)"
    )

    const [shader] = compileShader(v)

    expect(shader.vertexShader).toMatchSnapshot()
    expect(shader.fragmentShader).toMatchSnapshot()
  })

  it("resolves dependencies to other variables", () => {
    const float = Variable("float", 1)
    const root = Variable("float", float)

    const [shader] = compileShader(root)

    expect(shader.vertexShader).toMatchSnapshot()
  })

  it("doesn't render the same dependency twice", () => {
    const a = Float(1)
    const b = Bool(true, { inputs: { a, b: a } })

    const [shader] = compileShader(b)

    expect(shader.fragmentShader).toMatchSnapshot()
  })
})
