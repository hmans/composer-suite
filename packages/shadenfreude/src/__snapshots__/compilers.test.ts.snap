// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`compileShader doesn't render the same dependency twice 1`] = `
"void main()
{
  /*** BEGIN: Anonymous float = 1.0 (1) ***/
  float float_Anonymous_float_1_0_1;
  {
    float value = 1.0;
    float_Anonymous_float_1_0_1 = value;
  }
  /*** END: Anonymous float = 1.0 (1) ***/

  /*** BEGIN: Anonymous bool = true (2) ***/
  bool bool_Anonymous_bool_true_2;
  {
    float a = float_Anonymous_float_1_0_1;
    float b = float_Anonymous_float_1_0_1;
    bool value = true;
    bool_Anonymous_bool_true_2 = value;
  }
  /*** END: Anonymous bool = true (2) ***/

}"
`;

exports[`compileShader includes the variable's chunks if it has them 1`] = `
"/*** BEGIN: Anonymous float = 1.0 (1) ***/
uniform float u_time;
/*** END: Anonymous float = 1.0 (1) ***/

void main()
{
  /*** BEGIN: Anonymous float = 1.0 (1) ***/
  float float_Anonymous_float_1_0_1;
  {
    float value = 1.0;
    gl_Position = vec4(sin(u_time), 0.0, 0.0, 1.0);
    float_Anonymous_float_1_0_1 = value;
  }
  /*** END: Anonymous float = 1.0 (1) ***/

}"
`;

exports[`compileShader includes the variable's chunks if it has them 2`] = `
"/*** BEGIN: Anonymous float = 1.0 (1) ***/
uniform float u_time;
/*** END: Anonymous float = 1.0 (1) ***/

void main()
{
  /*** BEGIN: Anonymous float = 1.0 (1) ***/
  float float_Anonymous_float_1_0_1;
  {
    float value = 1.0;
    gl_FragColor = vec4(cos(u_time), 0.0, 0.0, 1.0);
    float_Anonymous_float_1_0_1 = value;
  }
  /*** END: Anonymous float = 1.0 (1) ***/

}"
`;

exports[`compileShader resolves dependencies to other variables 1`] = `
"void main()
{
  /*** BEGIN: Anonymous float = 1.0 (1) ***/
  float float_Anonymous_float_1_0_1;
  {
    float value = 1.0;
    float_Anonymous_float_1_0_1 = value;
  }
  /*** END: Anonymous float = 1.0 (1) ***/

  /*** BEGIN: Anonymous float = anonymous_4 (2) ***/
  float float_Anonymous_float_anonymous_4_2;
  {
    float value = float_Anonymous_float_1_0_1;
    float_Anonymous_float_anonymous_4_2 = value;
  }
  /*** END: Anonymous float = anonymous_4 (2) ***/

}"
`;

exports[`compileShader returns a fragmentShader 1`] = `
"void main()
{
  /*** BEGIN: Anonymous float = 1.0 (1) ***/
  float float_Anonymous_float_1_0_1;
  {
    float value = 1.0;
    float_Anonymous_float_1_0_1 = value;
  }
  /*** END: Anonymous float = 1.0 (1) ***/

}"
`;

exports[`compileShader returns a vertexShader 1`] = `
"void main()
{
  /*** BEGIN: Anonymous float = 1.0 (1) ***/
  float float_Anonymous_float_1_0_1;
  {
    float value = 1.0;
    float_Anonymous_float_1_0_1 = value;
  }
  /*** END: Anonymous float = 1.0 (1) ***/

}"
`;
